### ISSUE TITLE
datasignal dsclk returns 1 to all e2 in same tick even if e2 was executed by something else

### ISSUE THREAD

**Issue by AlexALX** [wiremod/wire#1585]
While debug my os as always i found that dsClk returns 1 to all e2 in same tick even if e2 was executed by something else like input or something else. Here is short instruction of about what i mean:

1. Spawn one e2 with this code:
```lua
@name DS Check 1
@inputs Send SEnt:entity
@persist 
@trigger 

if (~Send & Send) {
    print("DS Check 1")
    dsSendDirect("TEST",SEnt,"Test Message")   
}
```

2. Spawn second e2 with this code:
```lua
@name DS Check 2
@outputs Ent:entity TrigMe

if (first()|duped()) {
    Ent = entity()   
}

if (dsClk()) {
    print("DS Check 2", dsGetString())
    TrigMe = !TrigMe
}
```

3. Spawn third e2 with this code:
```lua
@name DS Check 3
@inputs TrigMe

if (~TrigMe) {
    print("DS Check 3", dsClk(), dsGetString())      
}
```

4. Connect Send input on DS Check 1 to button.
5. Connect SEnt:entity on DS Check 1 to Ent:entity from DS Check 2.
6. Connect TrigMe input on DS Check 3 to TrigMe output from DS Check 2.
7. Press button.

If everything connected right you will get:

```lua
DS Check 1	
DS Check 2	Test Message	
DS Check 3	1	Test Message	
```

So basically on DS Check 3 it reports that it was also triggered by datasignal, but it was triggered by input only! I'm not sure if this is a feature or bug but still reporting about this.

**Comment by adosikas** [comment]
This is because the datasignals use a shared variable (`currentsignal`, set here: https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/datasignal.lua#L91).

Storing that value in the target E2 context instead of as a shared variable should fix it.

**Comment by Divran** [comment]
This shouldn't be possible because datasignals are executed on the next tick https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/datasignal.lua#L110

**Comment by Divran** [comment]
I've got a fix for this
```lua
-- old code:
e2function number dsClk()
    return currentsignal ~= nil and 1 or 0
end

-- new code:
e2function number dsClk()
    if not currentsignal then return 0 end
    return currentsignal.to == self.entity and 1 or 0
end
```
this means you will still be able to run `dsGetString()` or whatever but that shouldn't matter because you won't be doing that if dsClk returns 0 anyway.
I'll merge this tomorrow if no one has anything against it

**Comment by Divran** [comment]
`runOnTick` works the same way btw and has always worked the same way since the beginning
https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/gametick.lua#L40

so if you erase `if (dsClk()) {` in your 2nd E2 and add `runOnTick`, then `tickClk()` will return 1 in the 3rd E2.

**Comment by AbigailBuccaneer** [comment]
Yeah, these should definitely use `self.data`, I'm happy to review changes.

**Comment by adosikas** [comment]
While we are at refractoring, maybe have a single variable that stores trigger reason like "input"/"tick"/"duped", instead of flags for everything?. And a `clkType()` or whatever function that allows getting that var in E2.
